noir-projects/noir-contracts/contracts/fpc_contract/src/main.nr

```diff
@@ -16,61 +20,115 @@ contract FPC {
+    /// for the user (for refund note of the AA) and sets a public teardown function (in which the partial note
+    /// will be finalized and the fee will be paid to FPC in public),
+    /// 2. then the private and public functions of a tx get executed,
+    /// 3. at this point we know the tx fee so we can compute in the teardown function how much of AA the user needs
+    /// to pay to the FPC and how much of it will be refunded back (this is computed based on an exchange rate between
+    /// AA and fee juice). Now we finalize the refund note for the user and we send the fee to the FPC in public,
+    /// 4. the protocol deducts the actual fee denominated in fee juice from the FPC's balance.
+    ///
+    /// With this scheme a user has privately paid for the tx fee with an arbitrary AA (e.g. could be a stablecoin).
+    ///
+    /// ***Note:***
+    /// This flow allows us to pay for the tx with msg_sender's private balance of AA and hence msg_sender's identity
+    /// is not revealed. We do, however, reveal the `max_fee`.
     #[private]
-    fn fee_entrypoint_private(amount: Field, asset: AztecAddress, nonce: Field) {
+    fn fee_entrypoint_private(max_fee: Field, nonce: Field) {
         // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
-        let settings = storage.settings.read();
+        let config = storage.config.read();
 
-        assert(asset == settings.other_asset);
-
-        Token::at(asset).setup_refund(settings.admin, context.msg_sender(), amount, nonce).call(
+        Token::at(config.accepted_asset).setup_refund(context.msg_sender(), max_fee, nonce).call(
             &mut context,
         );
         context.set_as_fee_payer();
     }
 
+    /// Pays for the tx fee with msg_sender's public balance of accepted asset (AA). The maximum fee a user is willing
+    /// to pay is defined by `max_fee` and is denominated in AA.
+    ///
+    /// ## Overview
+    /// The refund flow works as follows:
+    /// 1. We pull the `max_fee` from the user's balance of the AA to this contract (requires setting an authwit),
+    /// 2. then the private and public functions of a tx get executed,
+    /// 3. at this point we know the tx fee so we can compute how much of AA the user needs to pay to FPC and how much
+    /// of it will be refunded back. Note that this is computed based on an exchange rate between AA and fee juice.
+    /// 4. the protocol deducts the actual fee denominated in fee juice from the FPC's balance.
     #[private]
-    fn fee_entrypoint_public(amount: Field, asset: AztecAddress, nonce: Field) {
-        FPC::at(context.this_address())
-            .prepare_fee(context.msg_sender(), amount, asset, nonce)
+    fn fee_entrypoint_public(max_fee: Field, nonce: Field) {
+        // TODO(PR #8022): Once PublicImmutable performs only 1 merkle proof here, we'll save ~4k gates
```

Comment:
Because currently it's not compatible with how we allocate slot by macros and in that PR it works only because there is no state variable after the shared immutable. The issue is with the size of the variable in storage, currently we just expect it to be the num members in the struct but for the optimization to work we also store a hash.